language/src/main.rs

```diff
@@ -65,93 +68,275 @@ impl HacspecErrorEmitter for Session {
+                            rustc_ast::ast::ModKind::Unloaded,
+                        ) => false,
+                        _ => true,
+                    })
+                    .collect(),
+                span,
+            }
+        }
+    };
+
+    /////////////////////////////////
+    // Start of actual translation //
+    /////////////////////////////////
+
+    let external_data = |imported_crates: &Vec<rustspec::Spanned<String>>| {
+        queries.global_ctxt().unwrap().peek_mut().enter(|tcx| {
+            hir_to_rustspec::retrieve_external_data(&compiler.session(), &tcx, imported_crates)
+        })
+    };
+
+    let krate =
+        match ast_to_rustspec::translate(&compiler.session(), &krate, &external_data, new_specials)
         {
             Ok(krate) => krate,
             Err(_) => {
                 compiler
                     .session()
-                    .err("found some Hacspec typechecking errors");
+                    .err("unable to translate to Hacspec due to out-of-language errors");
                 return Compilation::Stop;
             }
         };
-        let imported_crates = name_resolution::get_imported_crates(&krate);
-        let imported_crates = imported_crates
-            .into_iter()
-            .filter(|(x, _)| x != "hacspec_lib")
-            .map(|(x, _)| x)
-            .collect::<Vec<_>>();
-        println!(
-            " > Successfully typechecked{}",
-            if imported_crates.len() == 0 {
-                ".".to_string()
-            } else {
-                format!(
-                    ", assuming that the code in crates {} has also been Hacspec-typechecked",
-                    imported_crates.iter().format(", ")
-                )
-            }
-        );
 
-        match &self.output_file {
-            None => return Compilation::Stop,
-            Some(file) => match Path::new(file).extension().and_then(OsStr::to_str).unwrap() {
+    let krate = match name_resolution::resolve_crate(
+        &compiler.session(),
+        krate,
+        &external_data,
+        new_top_ctx,
+    ) {
+        Ok(krate) => krate,
+        Err(_) => {
+            compiler
+                .session()
+                .err("found some Hacspec name resolution errors");
+            return Compilation::Stop;
+        }
+    };
+
+    let krate = match typechecker::typecheck_program(&compiler.session(), &krate, new_top_ctx) {
+        Ok(krate) => krate,
+        Err(_) => {
+            compiler
+                .session()
+                .err("found some Hacspec typechecking errors");
+            return Compilation::Stop;
+        }
+    };
+    let imported_crates = name_resolution::get_imported_crates(&krate);
+    let imported_crates = imported_crates
+        .into_iter()
+        .filter(|(x, _)| x != "hacspec_lib")
+        .map(|(x, _)| x)
+        .collect::<Vec<_>>();
+    println!(
+        " > Successfully typechecked {}{}",
+        krate_path.clone(),
```

Comment:
I think this is where the naming issues are coming from. There are a couple things here. Let me try to lay it out the way I see this right now.

`krate_path` here is the name of the library `.rs` file that we are typechecking. So this is *not* what we actually want to display here. What is being typechecked is the crate. So while the name here is correct for modules, it is not for the crate itself. This issue didn't exist before because only crate were typechecked.
With only one file per crate we also didn't have the issue of how to include a crate in a backend.

We have to define now how modules are supposed to work in hacspec and in the backends. Here's my proposal:
A crate get's extracted on a backend to a file with the crate name (not the library name you specify in cargo).
Modules within the crate get extracted with their module name as we do right now and included in the top level library file. This would be consistent with the way crates are used in Rust and doesn't require hacspec users to use specific names for their files (and would make the renaming of the examples unnecessary).
On the example of the `test-crate`: Instead of extracting a `Lib.v` we'd extract a `Testcrate.v`.

Reply:
So this is basically renaming the root file (often `lib.rs`) to the crate name when extracting. I am OK with this solution, however I think we should have a default export be the name of the crate, so it matches up with imports from other crates.

Reply:
> I think we should have a default export be the name of the crate

I'm not sure I fully understand what you mean. How is this different from using the crate name for the root file when extracting?

Reply:
It is not, I was just thinking we have a naming option for the root file as before, however the default is the crate name, rather than the file name or something else, for it to be consistent with imports. I have finished these changes, I think, we only need to revert the renaming of the crate root files.